### PR TITLE
Private Link Service Integration for k8s service

### LINKED ISSUE #872: Add support for Private Link Service when exposing service as LoadBalancer
#### What would you like to be added:

Add support for Private Link Service when exposing service as LoadBalancer with a new annotation `service.beta.kubernetes.io/azure-private-link-service: enabled`

Without this, customers have to perform the following steps manually:

step 1): ILB service creation: https://docs.microsoft.com/en-us/azure/aks/internal-lb#create-an-internal-load-balancer
step 2): find the ILB frontend configuration ID from above step and then follow https://docs.microsoft.com/en-us/azure/private-link/create-private-link-service-cli#create-a-private-link-service to create PLS.

#### Why is this needed:

Refer https://github.com/Azure/AKS/issues/1604.

### PR DESCRIPTION
**Pull request by jwtty**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #872 

#### Does this PR introduce a user-facing change?

```release-note
User can deploy managed Azure private link service resources by annotating the service object.
Detailed design doc can be found here: https://kubernetes-sigs.github.io/cloud-provider-azure/development/design-docs/pls-integration/
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MartinForReal** [comment]
Could you please move files under azureclients to another commit?

**Comment by feiskyer** [comment]
> Coverage decreased (-2.4%) to 77.688% when pulling https://github.com/kubernetes-sigs/cloud-provider-azure/commit/227b841ce0fdc3039450579596e159e9de4fbb17 on jwtty:pls into https://github.com/kubernetes-sigs/cloud-provider-azure/commit/2fcf748e8fb56460c6ea3514ad4c5f9f002e1950 on kubernetes-sigs:master.

Could you add some unit tests to ensure the unit test coverage?

File: pkg/provider/azure_privatelinkservice.go [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+package provider
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	v1 "k8s.io/api/core/v1"
```
Comment (nilo19) [inline pkg/provider/azure_privatelinkservice.go:26]: nit: categorize the imports

File: pkg/provider/azure_privatelinkservice.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+		if err != nil {
+			klog.Errorf("reconcilePrivateLinkService for service(%s): getPrivateLinkService(%s) failed: %v", serviceName, to.String(fipConfigID), err)
+			return err
+		}
+
+		exists := !strings.EqualFold(to.String(existingPLS.ID), consts.PrivateLinkServiceNotExistID)
+		if exists {
+			if !isManagedPrivateLinkSerivce(&existingPLS, clusterName) {
+				return fmt.Errorf(
+					"reconcilePrivateLinkService for service(%s) failed: LB frontend(%s) already has not managed private link service(%s)",
```
Comment (nilo19) [inline pkg/provider/azure_privatelinkservice.go:62]: not managed -> unmanaged?

File: pkg/provider/azure_privatelinkservice.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+					serviceName,
+					to.String(fipConfigID),
+					to.String(existingPLS.ID),
+				)
+			}
+			// If there is an existing private link service, only owner service can update its properties
+			ownerService := getPrivateLinkServiceOwner(&existingPLS)
+			if ownerService != "" && !strings.EqualFold(ownerService, serviceName) && serviceHasAdditionalConfigs(service) {
```
Comment (nilo19) [inline pkg/provider/azure_privatelinkservice.go:70]: is `ownerService != ""` nesessary?

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-cloud-provider-azure-e2e-ccm-vmss
--


<br class="Apple-interchange-newline">
</body>
</html>

File: pkg/provider/azure_loadbalancer.go [hunk 4]
```diff
@@ -94,7 +94,7 @@ func (az *Cloud) reconcileService(ctx context.Context, clusterName string, servi
 		return nil, err
 	}
 
-	lbStatus, _, err := az.getServiceLoadBalancerStatus(service, lb, nil)
+	lbStatus, fipConfig, err := az.getServiceLoadBalancerStatus(service, lb, nil)
```
[the remark this answers is not in the document]
Comment (jwtty) [reply pkg/provider/azure_loadbalancer.go:97]: 1. We actually only support PLS with internal LB so no public ip in the scenario.
2. If a svc is newly created, public IP is created when LB and frontend ip config are created in `reconcileLoadBalancer`: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/8d8ea5d4d4ad0e3a5c1e7ebc46ce99f7c32c2a52/pkg/provider/azure_loadbalancer.go#L1766

[the remark this answers is not in the document]
Comment (feiskyer) [reply pkg/provider/azure_loadbalancer.go:97]: by the way, we have already validated PLS on external LB would not work.

File: pkg/provider/azure_privatelinkservice.go [hunk 5]
```diff
@@ -0,0 +1,682 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (feiskyer) [inline pkg/provider/azure_privatelinkservice.go:2]: s/2020/2022

File: pkg/provider/azure_privatelinkservice.go [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+	service *v1.Service,
+	fipConfig *network.FrontendIPConfiguration,
+) (dirtyPLS bool, err error) {
+	dirtyPLS = false
+
+	if existingPLS == nil {
+		return false, fmt.Errorf("getExpectedPrivateLinkService: existingPLS is nil (unexpected)")
+	}
+
+	if existingPLS.Name == nil || !strings.EqualFold(*existingPLS.Name, *plsName) {
```
Comment (feiskyer) [inline pkg/provider/azure_privatelinkservice.go:245]: if a PLS is in a failed state, could we also set dirtyPLS to trigger a reconciling loop again?

**Comment by feiskyer** [comment]
e2e tests are failing due to quota limitations. let's run them again
/retest

File: pkg/provider/azure_privatelinkservice.go (hunk 5 above)
Comment (jwtty) [reply pkg/provider/azure_privatelinkservice.go:2]: Thanks, changing the test file also.

File: pkg/provider/azure_privatelinkservice.go (hunk 6 above)
Comment (jwtty) [reply pkg/provider/azure_privatelinkservice.go:245]: Sure, thanks!

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-cloud-provider-azure-e2e-ccm-vmss
--


<br class="Apple-interchange-newline">
</body>
</html>

**Pull request merged**